### PR TITLE
Fix handling of schemas with `.transform` in them

### LINKED ISSUE #31: Problem with `parse` method when `zod.transform` is used in a schema
The problem occurs when `transform` has been used in the schema.

I don't know what the intention was, but when parsing the schema, `zod class` performs double parsing. In the `parse` function(https://github.com/sam-goodwin/zod-class/blob/main/src/index.ts#L305) and in the constructor(https://github.com/sam-goodwin/zod-class/blob/main/src/index.ts#L187).

I think this is also an performance issue, because why would we parse the schema twice?

There is also a problem of types in constructor when we use `transform` from `zod`.

Example code to reproduce problem:

https://stackblitz.com/edit/stackblitz-starters-cg4bsqz2?file=main.ts&view=editor

```typescript
import { z } from 'zod';
import { Z } from 'zod-class';

enum Color {
  Red,
  Blue,
  Green,
}

enum ColorAsString {
  Red = 'red',
  Blue = 'blue',
  Green = 'green',
}

const ColorMap = {
  [ColorAsString.Red]: Color.Red,
  [ColorAsString.Blue]: Color.Blue,
  [ColorAsString.Green]: Color.Green,
} satisfies Record<ColorAsString, Color>;

class CustomSchema extends Z.class({
  color: z.nativeEnum(ColorAsString).transform((c) => ColorMap[c]),
}) {}

// Works fine, but types are incorrect
console.log(new CustomSchema({ color: ColorAsString.Blue }));

// Throws an exception because it parses twice
console.log(CustomSchema.parse({ color: ColorAsString.Blue }));
```

### PR DESCRIPTION
**Pull request by netroy**
Fixes #31

### CODE REVIEW

File: test/zod-class.test.ts [hunk 1]
```diff
@@ -472,3 +470,26 @@ test("optional object fields are optional", () => {
...
+// see: https://github.com/sam-goodwin/zod-class/issues/31
+test.only("should support transform in schema", async () => {
+  class Test extends Z.class({
+    prop: z.enum(['true', 'false']).transform((val) => val === 'true'),
+  }) {}
+
+  const input = { prop: 'true' } as const;
+
+  // This should work and not have any type errors
+  const instance = new Test(input);
```
Comment (despairblue) [inline test/zod-class.test.ts:483]: Vice versa this should be a type error I think?

```ts
  new Test({ prop: true })
```

**Comment by despairblue** [comment]
I added some suggestions here: https://github.com/netroy/zod-class/pull/1

File: src/index.ts [hunk 2]
```diff
@@ -121,45 +121,85 @@ export interface ZodClass<
...
+    ? Z.output<T>
     : T extends ZodPromise<infer T>
-    ? Promise<Z.infer<T>>
+    ? Promise<Z.output<T>>
     : T extends ZodType<any, any, any>
-    ? z.infer<T>
+    ? z.output<T>
+    : never;
+
+  export type input<T> = T extends new (...args: any[]) => infer R
```
Comment (sam-goodwin) [inline src/index.ts:168]: Curious, what was the reason to separate out into input/output?

**Review by sam-goodwin — APPROVED** [review]

**Comment by sam-goodwin** [comment]
Thanks @netroy and @despairblue, the fix is released in `0.0.16 `

File: src/index.ts (hunk 2 above)
Comment (despairblue) [reply src/index.ts:168]: The constructors argument should be of the type `z.input<schema>` and the instance type of the class should be `z.output<schema>` with `z.input` being the type of what should go into `z.parse` and `z.output` is the type it returns if parsed successfully which includes all transforms which can change the type.

I could try to parameterize this type and have `z.input` and `z.output` passed in as a type parameter to reduce the duplication. Should I give that a try?

I hope I got the question right, if not let me know :)

**Pull request merged**